Added the default newsletter

tpatel opened the pull request:
refs https://github.com/TryGhost/Team/issues/1473

- We use the title to populate the newsletter same, slug and sender name
- We use the description to populate the newsletter description
- We use the global design settings to populate the corresponding newsletter design settings

I've enabled using a for-loop in this migration because it's iterating on an array of 9 elements which seems fine.

tpatel committed:
Added the default newsletter

refs https://github.com/TryGhost/Team/issues/1473

- We use the title to populate the newsletter same, slug and sender name
- We use the description to populate the newsletter description
- We use the global design settings to populate the corresponding newsletter design settings

tpatel commented:
I've been hitting a [weird error](https://github.com/TryGhost/Ghost/runs/5992196380?check_suite_focus=true) :

- The “Update from latest v3” action is failing with `A GhostError occurred.` and `A ProcessError occurred.`
- The “Print debug logs” action is failing with binary operator expected

I couldn't reproduce the issue locally when booting Ghost `3.42.9` and then migrating to `4.44` using `knex-migrator`.

daniellockyer commented:
@tpatel 

I think it's due to:

```
[10:35:53] → Ghost was able to start, but errored during boot with: `sqlite` does not support inserting default values. Specify values explicitly or use the `useNullAsDefault` config flag. (see docs http://knexjs.org/#Builder-insert).

```

tpatel commented:
@daniellockyer Thanks, I haven't found what exactly is trying to insert a default value, I'm gonna look into that again

tpatel committed:
Testing the `sender_email` value

This commit shouldn't be merged, it's only to test on the CI as I can't reproduce the issue locally

tpatel committed:
revert CI test + add logging

tpatel committed:
make sure this is the newsletter insert that's breaking the build

tpatel committed:
Attempt to prevent an edge case where the migration runs before the settings are populated

This is a commit as I cannot reproduce the issue locally

tpatel committed:
fix lint errors

matthanley commented on core/server/data/migrations/versions/4.44/2022-04-11-12-13-add-default-newsletter.js:
```diff
@@ -0,0 +1,84 @@
+
+            if (setting) {
+                return setting.value;
+            } else if (Object.hasOwnProperty.call(defaultSettings, name)) {
```
Minor point, but this implementation may be difficult to reason about because it can take a default value from two different sources.

tpatel replied:
I've rewritten everything, I think it's simpler now

matthanley commented on core/server/data/migrations/versions/4.44/2022-04-11-12-13-add-default-newsletter.js:
```diff
@@ -0,0 +1,84 @@
+        logging.info('Adding the default newsletter');
+
+        logging.info(knex('newsletters').insert(newsletter).toSQL().sql);
+        logging.info(JSON.stringify(knex('newsletters').insert(newsletter).toSQL().bindings));
```
I guess these additional log lines are left in from debugging?

tpatel replied:
good catch 👍

rishabhgrg commented on core/server/data/migrations/versions/4.44/2022-04-11-12-13-add-default-newsletter.js:
```diff
@@ -0,0 +1,84 @@
+        const designSettings = ['header_image', 'show_header_icon', 'show_header_title', 'title_font_category', 'title_alignment', 'show_feature_image', 'body_font_category', 'footer_content', 'show_badge'];
+        // eslint-disable-next-line no-restricted-syntax
+        for (const setting of designSettings) {
+            let value = await getSetting('newsletter_' + setting);
```
@tpatel I am wondering if its probably better to load all the desired setting from DB with a single knex query rather than fetching each setting individually ? Something like - 
```
await knex('settings')
            .whereIn('key', allSettings)
            .select(['key', 'value'])
```

which will fetch all settings from DB with a single query and we can then just use them directly after that ?

tpatel replied:
I've rewritten so that it's only one sql query to get all settings.

tpatel committed:
group all setting fetches into one sql query

rishabhgrg commented on core/server/data/migrations/versions/4.44/2022-04-11-12-13-add-default-newsletter.js:
```diff
@@ -0,0 +1,87 @@
+            id: ObjectId().toHexString(),
+            name: settingsMap.title,
+            description: settingsMap.description,
+            slug: slugify(settingsMap.title),
```
I think this should be `default-newsletter`, because that is the slug we are giving to default fixture as well and we don't allow slugs to be changed at the moment. Since there will be no newsletter in DB when this is run, its better to have a consistent slug for default newsletter across fixtures/migration, wdyt ?

tpatel replied:
fixed 👍

matthanley commented on core/server/data/migrations/versions/4.44/2022-04-11-12-13-add-default-newsletter.js:
```diff
@@ -0,0 +1,87 @@
+            name: settingsMap.title,
+            description: settingsMap.description,
+            slug: slugify(settingsMap.title),
+            sender_name: settingsMap.title,
```
I think this should probably be `null` after https://github.com/TryGhost/Ghost/pull/14479 to maintain current behaviour

tpatel replied:
ok I'll wait for this other PR to get merged before moving forward with this one

rishabhgrg replied:
@tpatel Merged the other PR now so this is good to be updated :)

tpatel replied:
awesome, I've updated the slug and sender_name changes based on your PR 👍

tpatel committed:
updated the default `slug` and `sender_name` to match the fixture from https://github.com/TryGhost/Ghost/pull/14479

tpatel committed:
fixed linting issue

tpatel committed:
renamed migration so that it happens after the sender_name migration

matthanley reviewed (commented):
tbh this migration feels difficult to follow/unnecessarily complex, as if it's had new bits added to it without rethinking the whole approach.

I've taken the liberty of reimplementing it [here](https://github.com/matthanley/Ghost/blob/8e1adf003d91c950ccc7e300aceccdfa94706ebb/core/server/data/migrations/versions/4.44/2022-04-13-13-00-add-default-newsletter.js) as an example of how I'd approach it. Let me know what you think!

tpatel committed:
switched to Matt's implementation

- removed loading the full lodash library in favor for only loading the `startsWith` library

tpatel commented:
I usually do prefer having the configuration separated from the code which is why it was structured that way. I also prefer using regexp as it avoids depending on lodash. I'm not attached to my implementation so I'm fine with changing it.

tpatel commented:
Just a note: this is only adding the default newsletter and not subscribing the existing members to this newsletter as this is part of [another task](https://github.com/TryGhost/Team/issues/1508) that is planned later in the feature development.

tpatel committed:
Moved the migration to the next Ghost version

matthanley commented:
We should hold off merging this and merge everything at once - otherwise if people change their design settings those changes will get reverted when we release newsletters

daniellockyer commented on core/server/data/migrations/versions/4.45/2022-04-13-13-00-add-default-newsletter.js:
```diff
@@ -0,0 +1,82 @@
+    },
+    async function down() {
+        // Noop because we don't want to reset the default newsletter values
+        logging.info(`Skipping newsletter design settings backfill rollack - not needed`);
```
`s/rollack/rollback/`

tpatel committed:
fixed typo

matthanley committed:
Updated migration version

matthanley committed:
Added show_header_name

SimonBackx commented:
We might want to hold off on merging this until I added the `created_at` and `updated_at` timestamps to the newsletters table
